### PR TITLE
Add table of contents dropdown list

### PR DESCRIPTION
**Pull request by MSakamaki**
I have add #985  features.

### CODE REVIEW

**Comment by rviscomi** [comment]
Based on the screenshots, my only two comments are about the repeated "Table of Contents" option and the divider separating each part.

image

Should we rename one of them? Open to ideas.

image

Could we remove the divider and see how it looks?

**Comment by MSakamaki** [comment]
@rviscomi 
Okay! I attaching a screenshot of the design with the divider removed and a screenshot of the old design. (I'm sorry, divider may have been overdone)
hr-diffine

I wanted to provide the same UX as the other dropdown list, so I thinking.
Luckily we had the translated word "close", so we thought it would be natural to set it this.
and, I thought about "Close the Table of Contents" but it was too long and broke the layout, so I opted for just "Close".
dropdown-opened


There is still a bit of behavior in the click event that bothers me, so I will update the PR as soon as it is resolved.

**Comment by MSakamaki** [comment]
Thanks for the great thoughts, @bazzadp !
as you pointed out, there were a number of oversights.
I'm going for the fixed!

**Comment by bazzadp** [comment]
As for the linting errors annoyingly it doesn't like using `{% if ... %}` statements inside HTML elements so instead of this:

```html
    <option 
        {% if current_title == title %} selected {% endif %}
        value="{{ url_for('chapter', year=year, lang=lang, chapter=chapter_config.slug) }}">
        {{ self.chapter() }} {{ chapter_config.chapter }}: {{title}}
      </option>
```

You have to do this:

```html
     {% if current_title == title %}
     <option selected
        value="{{ url_for('chapter', year=year, lang=lang, chapter=chapter_config.slug) }}">
        {{ self.chapter() }} {{ chapter_config.chapter }}: {{title}}
      </option>
      {% else %}
     <option
        value="{{ url_for('chapter', year=year, lang=lang, chapter=chapter_config.slug) }}">
        {{ self.chapter() }} {{ chapter_config.chapter }}: {{title}}
      </option>
      {% endif %}
```

It's a bit verbose and repetitive, but only way I've found of getting it to pass the linting unfortunately.

**Comment by bazzadp** [comment]
> * I think Rick's Main issue is with having the ToC button and the ToC label right after each other. What about moving the ToC to the bottom in an Appendices section instead of being at the top to resolve that? I think i'd prefer that to changing the title to Close.

Or alternatively, leave it where it is and change the label to Foreword? As the ToC page contains a Foreword as well and maybe it should be before the chapters.

WDYT @rviscomi @MSakamaki ?

**Comment by MSakamaki** [comment]
@bazzadp 
- Remove consecutive table of contents labels(add a Foreword)
- Added a table of contents and contributors to the Appendices section
- Add e-book links.
- Removing the horizon line
- Add a link for when javascript is disabled
- Hide the table of contents when it's 2020.

> Should we disable the current page? For example, if on CSS chapter, then disable CSS link?

I have not yet implemented link disabling for the current page due to the following points

- I more concerned about the problem of source code complexity than the value gained from the fix.
- I couldn't figure out how to take the current page for the table of contents, contributors, and methodology pages.

File: src/templates/base/2019/base.html
Comment (bazzadp) [inline src/templates/base/2019/base.html:163]: I found a why to handle the currently selected item. It's a little extra code but think worth is as you do it for the chapters:

```suggestion
    {% if request.path.endswith("methodology") %}
    <option selected disabled value="{{ url_for('methodology', year=year, lang=lang) }}">
      {{ self.methodology_title() }}
    </option>
    {% else %}
    <option value="{{ url_for('methodology', year=year, lang=lang) }}">
      {{ self.methodology_title() }}
    </option>
    {% endif %}
    {% if request.path.endswith("contributors") %}
    <option selected disabled value="{{ url_for('contributors', year=year, lang=lang) }}">
      {{ self.contributors_title() }}
    </option>
    {% else %}
    <option value="{{ url_for('contributors', year=year, lang=lang) }}">
      {{ self.contributors_title() }}
    </option>
    {% endif %}
```

File: src/templates/base/2019/base.html
Comment (bazzadp) [inline src/templates/base/2019/base.html:181]: The foreword looks a little lonely without a heading. What do you think about this:

```suggestion
    <li class="nav-dropdown-list__part">
      {{ self.introduction() }}
    </li>
    <li class="nav-dropdown-list__chapter">
      {% if request.path.endswith("table-of-contents") %}
      <span class="nav-dropdown-list__current">{{ self.foreword_title() }}</span>
      {% else %}
      <a href="{{ url_for('table_of_contents', year=year, lang=lang) }}">{{ self.foreword_title() }}</a>
      {% endif %}
    </li>
```

Addition of Introduction header in menu

File: src/templates/base/2019/base.html
Comment (bazzadp) [inline src/templates/base/2019/base.html:189]: Suggesting the newly added `nav-dropdown-list__current` class here here to grey it out (though could also be renamed active if you prefer):

```suggestion
    <li class="nav-dropdown-list__chapter {% if current_title == title %}nav-dropdown-list__current{% endif %}">
```

File: src/templates/base/2019/base.html
Comment (bazzadp) [inline src/templates/base/2019/base.html:215]: Grey out if current page:

```suggestion
    <li class="nav-dropdown-list__chapter">
      {% if request.path.endswith("methodology") %}
      <span class="nav-dropdown-list__current">{{ self.methodology_title() }}</span>
      {% else %}
      <a href="{{ url_for('methodology', year=year, lang=lang) }}">{{ self.methodology_title() }}</a>
      {% endif %}
    </li>
    <li class="nav-dropdown-list__chapter">
      {% if request.path.endswith("contributors") %}
      <span class="nav-dropdown-list__current">{{ self.contributors_title() }}</span>
      {% else %}
      <a href="{{ url_for('contributors', year=year, lang=lang) }}">{{ self.contributors_title() }}</a>
      {% endif %}
    </li>
```

File: src/templates/base/2019/base.html
Comment (bazzadp) [inline src/templates/base/2019/base.html:122]: Forgot to update this so have duplicate ids at the moment:

```suggestion
  <select id="table-of-contents-switcher-{{switcher_name}}">
```

**Review by bazzadp — CHANGES_REQUESTED** [review]
> I have not yet implemented link disabling for the current page due to the following points
> 
> I more concerned about the problem of source code complexity than the value gained from the fix.
> I couldn't figure out how to take the current page for the table of contents, contributors, and methodology pages.
> 

I take the point about source code complexity but then we shouldn't leave it in for the chapters as currently it's half in and half out.

However I think I've found a way to implement this with only a little extra code. See some of my suggested changes.

> And subsequently resolve the linting error.
> (If way to run the lint locally, I helpful to let me know.

Haven't found an easy way to do this myself to be honest! However if you [enable GitHub Actions on your fork](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/disabling-or-limiting-github-actions-for-a-repository) then it will lint on each push up there, before you open the Pull Request so will at least give you heads up before you open the PR.

File: src/templates/base/2019/base.html
Comment (rviscomi) [reply src/templates/base/2019/base.html:181]: How about "Foreword" itself being the header?

Comment (bazzadp) [reply src/templates/base/2019/base.html:181]: And what as the item? "Table of Contents"? Don't like the idead of making the header clickable as others aren't so need an item IMHO.

**Comment by rviscomi** [comment]
I'll defer to @bazzadp to finalize this review. Thanks again for working on this @MSakamaki!

File: src/templates/base/2019/base.html
Comment (rviscomi) [reply src/templates/base/2019/base.html:181]: No item IMO. Just as it's ok to have an H2 with no H3 in the outline sidebar, this component could have headings with no items.

Comment (bazzadp) [reply src/templates/base/2019/base.html:181]: OK but then the heading will be a link. But "Part 1. Page content" will not be a link - and neither will any of the other headers. Isn't that inconsistent and confusing?

Comment (rviscomi) [reply src/templates/base/2019/base.html:181]: We could make Part X headings link to the first chapter in their section too for consistency. Otherwise I think its ok.

Comment (bazzadp) [reply src/templates/base/2019/base.html:181]: That's not a bad idea. Wasn't that happy with the headings not being links as kind of similar in style to the items.

Could also link to here: https://almanac.httparchive.org/en/2019/table-of-contents#part-1

Will have a think. Let's come back to this one and address the other issues first.

Comment (rviscomi) [reply src/templates/base/2019/base.html:181]: Yeah deep linking to the part in the TOC sounds like a good idea to me. 👍

**Comment by rviscomi** [comment]
Any progress on this @MSakamaki? Hoping to resolve all PRs this week.

**Comment by rviscomi** [comment]
Hi @MSakamaki let us know if you need help getting this PR over the finish line :) 🏁

**Comment by MSakamaki** [comment]
@rviscomi 
sorry for the delay!
I'll fix this this one this week now that I've had time to do so!

File: src/static/css/2019.css
Comment (MSakamaki) [reply src/static/css/2019.css:1364]: fixed.

File: src/static/css/2019.css
Comment (MSakamaki) [reply src/static/css/2019.css:1378]: fixed.

File: src/static/css/2019.css
Comment (MSakamaki) [reply src/static/css/2019.css:1368]: fixed.

File: src/static/css/2019.css
Comment (MSakamaki) [reply src/static/css/2019.css:1374]: fixed.

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:124]: ok i fixed !

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:135]: i fixed.

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:145]: i fixed.

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:163]: this is the solution I've been looking for. Thank you!

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:181]: ok i fixed.

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:189]: i fixed.

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:215]: i fixed.
I fixed part deep links.
https://github.com/HTTPArchive/almanac.httparchive.org/pull/1393#discussion_r516162235

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html:122]: i fixed.

File: src/server/routes.py
Comment (MSakamaki) [reply src/server/routes.py:59]: i fixed.

File: src/static/css/2019.css
Comment (MSakamaki) [reply src/static/css/2019.css]: i fixed.

**Review by MSakamaki — COMMENTED** [review]
@bazzadp 
> And subsequently resolve the linting error.
> (If way to run the lint locally, I helpful to let me know.

I've found a way to do this locally during development.
Here's what the command looks like.

```sh
# first only
docker pull github/super-linter:v3.13.2

# run linting
docker run -e RUN_LOCAL=true \
         -e VALIDATE_ALL_CODEBASE=false \
         -e VALIDATE_BASH=true \
         -e VALIDATE_CSS=true \
         -e VALIDATE_HTML=true \
         -e VALIDATE_JAVASCRIPT_ES=true \
         -e VALIDATE_JSON=true \
         -e VALIDATE_MD=true \
         -e VALIDATE_PYTHON_PYLINT=true \
         -e VALIDATE_PYTHON_FLAKE8=true \
         -e VALIDATE_YAML=true \
         -e FILTER_REGEX_EXCLUDE=".*(github|node_modules).*" \
         -v /Users/m.sakamaki/project/sakamaki/almanac.httparchive.org:/tmp/lint \
         github/super-linter
```
* Inherited from .github/workflows/linter.yml

this [document](https://github.com/github/super-linter/blob/master/docs/run-linter-locally.md) contains instructions for running super-linter it locally.
can I add instructions for doing the super-linter locally to README.md?
If there are no problems, a create issues.

Note: I tried, but the node_modules folder cannot be excluded by FILTER_REGEX_EXCLUDE, so a little research is needed. (Deleting the folder will make it go faster.)

https://github.com/github/super-linter/issues/710

**Comment by bazzadp** [comment]
> @bazzadp
> 
> > And subsequently resolve the linting error.
> > (If way to run the lint locally, I helpful to let me know.
> 
> I've found a way to do this locally during development.
> Here's what the command looks like.
> 
> ```shell
> # first only
> docker pull github/super-linter:v3.13.2
> 
> # run linting
> docker run -e RUN_LOCAL=true \
>          -e VALIDATE_ALL_CODEBASE=false \
>          -e VALIDATE_BASH=true \
>          -e VALIDATE_CSS=true \
>          -e VALIDATE_HTML=true \
>          -e VALIDATE_JAVASCRIPT_ES=true \
>          -e VALIDATE_JSON=true \
>          -e VALIDATE_MD=true \
>          -e VALIDATE_PYTHON_PYLINT=true \
>          -e VALIDATE_PYTHON_FLAKE8=true \
>          -e VALIDATE_YAML=true \
>          -e FILTER_REGEX_EXCLUDE=".*(github|node_modules).*" \
>          -v [project directory]:/tmp/lint \
>          github/super-linter
> ```
> 
> * Inherited from .github/workflows/linter.yml
> 
> this [document](https://github.com/github/super-linter/blob/master/docs/run-linter-locally.md) contains instructions for running super-linter it locally.
> can I add instructions for doing the super-linter locally to README.md?
> If there are no problems, a create issues.
> 
> Note: I tried, but the node_modules folder cannot be excluded by FILTER_REGEX_EXCLUDE, so a little research is needed. (Deleting the folder will make it go faster.)
> 
> [github/super-linter#710](https://github.com/github/super-linter/issues/710)

Oh cool - yes please do add this!

**Comment by bazzadp** [comment]
Is this ready for another review yet @MSakamaki ?

BTW there have been quite a few changes to `main` since this was opened so think it would be good to merge `main` into this branch to ensure there are no issues.

**Comment by MSakamaki** [comment]
@bazzadp Yes, I'm fine!
I'll be following the main branch now!

File: src/server/routes.py
Comment (bazzadp) [reply src/server/routes.py:59]: It doesn't look like this was fixed @MSakamaki

File: src/templates/base/2019/base.html
Comment (bazzadp) [reply src/templates/base/2019/base.html]: Doesn't look like this was fixed

File: src/static/css/2019.css
Comment (bazzadp) [inline src/static/css/2019.css:1366]: Add some extra padding on the left:

```suggestion
  padding: 16px 8px 8px 16px;
```

File: src/static/css/2019.css
Comment (bazzadp) [inline src/static/css/2019.css:1381]: Extra padding as we've padded the part headers too:

```suggestion
  padding: 8px 8px 8px 24px;
```

**Review by bazzadp — CHANGES_REQUESTED** [review]
I think we're almost there @MSakamaki !

Some more changes first though.

File: src/server/routes.py
Comment (MSakamaki) [reply src/server/routes.py:59]: I'm Sorry, after merging into main, the table of contents was not displayed because the DEFAUL_YEAR was set to 2020.

It is now shown in the next fix.
https://github.com/HTTPArchive/almanac.httparchive.org/pull/1393/files#r521649741 

are you okay with this?
2020-accessibility-statement

File: src/templates/base/2019/base.html
Comment (MSakamaki) [reply src/templates/base/2019/base.html]: i fixed.

File: src/templates/base/2019/base.html
Comment (bazzadp) [reply src/templates/base/2019/base.html]: The title is still set to `self.foreword_title()` instead of `self.table_of_contents_title()`. As I say on Mobile I think we should use `self.table_of_contents_title()` as that's what shows on home page for example.

File: src/static/css/2019.css
Comment (bazzadp) [reply src/static/css/2019.css]: Yeah that's an intentional reverse. It's a little bit odd but think the best way as it's going up, instead of the usual down.

Will resolve this now. Thanks!

File: src/templates/base/2019/base.html
Comment (bazzadp) [inline src/templates/base/2019/base.html:121]:
```suggestion
    <option value="{{ url_for('table_of_contents', year=year, lang=lang) }}">{{ self.table_of_contents_title() }}</option>
```

**Review by bazzadp — APPROVED** [review]
Just one small thing to change. Will commit that and then merge this.

Thanks so much for working on this @MSakamaki - looks awesome and will be super useful!

File: src/templates/base/2019/base.html
Comment (bazzadp) [inline src/templates/base/2019/base.html:121]: Actually I'm going to do one more change here as can't select ToC on mobile on home page (as that's the current page):

```suggestion
    <option value="{{ url_for('table_of_contents', year=year, lang=lang) }}">{{ self.table_of_contents_title() }}</option>
    <option value="{{ url_for('table_of_contents', year=year, lang=lang) }}#forward">{{ self.foreword_title() }}</option>
```

**Pull request merged**
